HARK/core.py

```diff
@@ -18,11 +19,201 @@
+        self._term_age = parameters.get("T_cycle", None)
+        self._age_inv = []
+        self._age_var = []
+
+        for key, value in parameters.items():
+            setattr(self, key, value)
+
+    def __infer_dims__(self, key, value):
+        """
+        Infers the age-varying dimensions of a parameter.
+
+        If the parameter is a scalar, numpy array, or None, it is assumed to be
+        invariant over time. If the parameter is a list or tuple, it is assumed
+        to be varying over time. If the parameter is a list or tuple of length
+        greater than 1, the length of the list or tuple must match the
+        `_term_age` attribute of the Parameters object.
+
+        Parameters
+        ----------
+        key : str
+            name of parameter
+        value : Any
+            value of parameter
+
+        """
+        if isinstance(value, (int, float, np.ndarray, type(None))):
+            self.__add_to_time_inv(key)
+            return value
+        if isinstance(value, (list, tuple)):
+            if len(value) == 1:
+                self.__add_to_time_inv(key)
+                return value[0]
+            if self._term_age is None or self._term_age == 1:
+                self._term_age = len(value)
+            if len(value) == self._term_age:
+                self.__add_to_time_vary(key)
+                return value
+            raise ValueError(f"Parameter {key} must be of length 1 or {self._term_age}")
+        raise ValueError(f"Parameter {key} has type {type(value)}")
+
+    def __add_to_time_inv(self, key):
+        """
+        Adds parameter name to invariant list and removes from varying list.
+
+        Parameters
+        ----------
+        key : str
+            parameter name
+        """
+        if key in self._age_var:
+            self._age_var.remove(key)
+        if key not in self._age_inv:
+            self._age_inv.append(key)
+
+    def __add_to_time_vary(self, key):
+        """
+        Adds parameter name to varying list and removes from invariant list.
+
+        Parameters
+        ----------
+        key : str
+            parameter name
+        """
+        if key in self._age_inv:
+            self._age_inv.remove(key)
+        if key not in self._age_var:
+            self._age_var.append(key)
+
+    def __getitem__(self, age_or_key):
+        """
+        If age_or_key is an integer, returns a Parameters object with the parameters
+        that apply to that age. This includes all invariant parameters and the
+        `age_or_key`th element of all age-varying parameters. If age_or_key is a string,
+        it returns the value of the parameter with that name.
+
+        Parameters
+        ----------
+        age_or_key : int or str
+            Age or key of parameter(s)
+
+
+        Returns
+        -------
+        Parameters or value
+            Parameters object with parameters that apply to age `age_or_key` or value of
+            parameter with name `age_or_key`.
```

Comment:
Maybe I'm missing something, but it feels a little odd for the return type of this in the case where the argument is an `age` to be a Parameters object.

The Parameters object seem to be about an Agent's entire parameterization, including the age of the agent...but what does it mean to be a Parameters object for one age slice of an agent?

Consider having the return type in the `age` case be a dictionary or named tuple.

Reply:
Yeah I just thought the slice of a Parameter should also be a Parameter. 

We can think of Parameter objects as inputs to agents, but also as inputs to Solvers, or Simulators, or Estimators. Sometimes these objects will only take one `age` at a time, but I thought they should still receive a collection of parameters wrapped in a Parameters object. Then if we actually need a primitive like a dictionary or named tuple, we could easily get that from the Parameters slice.

Reply:
I am wondering to what extent this object is a novel data structure, and to what extent it has specific semantics based on its uses in modeling.

In many ways, the fundamental data structure is an array. We might accomplish a lot of the same thing with a $T \times P$ array, where $T$ is the maximum age and $P$ is the number of parameters. Then we would 'slice' literally by time using the same well-known array slicing interface.

Of course, the difference in this case is that some parameter rows are being defined, up front, as scalars that do not vary. This is largely for convenience -- we don't want to have to manually write in long lists of constant value.

If `Parameters` is a data structure, then I suppose it makes sense for a slice of `Parameters` to be another `Parameters` object.

But if part of the semantics of `Parameters` is that it has to do with agent ages, then I wonder if the abstraction breaks down.

For example, consider:

```
In [1]: from HARK.core import Parameters

In [2]: from HARK.ConsumptionSaving.ConsIndShockModel import init_lifecycle

In [3]: params = Parameters(**init_lifecycle)

In [4]: params._term_age
Out[4]: 65

In [5]: param_slice = params[10]
```

What is `param_slice._term_age`?

The docs say that `_term_age` is "The terminal age of the agents in the model." But wouldn't `param_slice` be a Parameters object with no age varying parameters? Does `_term_age` maintain the original reference, or does it become 1?

IIRC, when you take a slice of a Python array, you wind up with a view into that array, rather than a copy of that array. This has consequences: when you mutate the slice, you mutate the original array. As in:

```
>>> a = np.zeros((3,3))
>>> b = a[1:2,1:2]
>>> b[0,0] = 3
>>> a
array([[0., 0., 0.],
       [0., 3., 0.],
       [0., 0., 0.]])
```

I don't think that with your current implementation, mutating slices would mutate the underlying Parameters object. That may or may not be desirable. But I hope these choices can be made in a considered way.

Reply:
The `_term_age `of a slice would be 1, as it re-interprets all the parameters given, which are now only for a particular time period. 

The intended use of this data structure is to both handle the age dimension of models and to solve the "white board" problem.